conveyor/slot_test.go

```diff
@@ -0,0 +1,255 @@
+/*
+ *    Copyright 2019 Insolar Technologies
+ *
+ *    Licensed under the Apache License, Version 2.0 (the "License");
+ *    you may not use this file except in compliance with the License.
+ *    You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+package conveyor
+
+import (
+	"testing"
+
+	"github.com/insolar/insolar/conveyor/queue"
+	"github.com/insolar/insolar/core"
+	"github.com/stretchr/testify/require"
+)
+
+func testSlot(t *testing.T, isQueueOk bool, pulseNumber core.PulseNumber) *Slot {
+	var q *queue.IQueueMock
+	if isQueueOk {
+		q = mockQueue(t)
+	} else {
+		q = mockQueueReturnFalse(t)
+	}
+
+	return &Slot{
+		inputQueue:  q,
+		pulseNumber: pulseNumber,
+		pulse:       core.Pulse{PulseNumber: pulseNumber},
+	}
+}
+
+func len3List() ElementList {
+	el1 := &slotElement{id: 1}
+	el2 := &slotElement{id: 2}
+	el3 := &slotElement{id: 3}
+	el1.listNext = el2
+	el2.listNext = el3
+
+	l := ElementList{
+		head: el1,
+		tail: el3,
+	}
+	return l
+}
+
+func TestElementList_len_Empty(t *testing.T) {
+	l := ElementList{}
+	listLen := l.len()
+	require.Equal(t, 0, listLen)
+}
+
+func TestElementList_len(t *testing.T) {
+	l := len3List()
+	listLen := l.len()
+	require.Equal(t, 3, listLen)
+}
+
+func TestElementList_popElement_Empty(t *testing.T) {
+	l := ElementList{}
+	el := l.popElement()
+	require.Nil(t, el)
+}
+
+func TestElementList_popElement_OnlyOne(t *testing.T) {
+	expectedElement := &slotElement{id: 1}
+	l := ElementList{head: expectedElement, tail: expectedElement}
+
+	el := l.popElement()
+	require.Equal(t, expectedElement, el)
+	require.Equal(t, 0, l.len())
+}
+
+func TestElementList_popElement(t *testing.T) {
+	l := len3List()
+	prevHead := l.head
+	prevTail := l.tail
+
+	el := l.popElement()
+	require.Equal(t, prevHead, el)
+	require.Equal(t, prevHead.listNext, l.head)
+	require.Equal(t, prevTail, l.tail)
+	require.Equal(t, 2, l.len())
+}
+
+func TestElementList_pushElement_Empty(t *testing.T) {
```

Comment:
don't see case, when we popElement multiple times

Reply:
and tests with alternations of popElement and pushElement
and test with multiple pushElement